src/audio/google_audio_post_processing.c

```diff
@@ -0,0 +1,551 @@
+	case SOF_GAPP_MODEL:
+		COMPILER_FALLTHROUGH
+	case SOF_GAPP_DATA:
+		return gapp_get_internal_config(cd, cdata, max_data_size);
+	default:
+		comp_err(dev,
+			 "gapp_ctrl_get_bin_data(): unknown binary data type");
+		return -EINVAL;
+	}
+}
+
+static int gapp_ctrl_get_data(struct comp_dev *dev,
+			      struct sof_ipc_ctrl_data *cdata,
+			      int max_data_size)
+{
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		return gapp_ctrl_get_bin_data(dev, cdata, max_data_size);
+	default:
+		comp_err(dev, "smart_amp_ctrl_get_data(): invalid cdata->cmd");
+		return -EINVAL;
+	}
+}
+
+static int gapp_cmd(struct comp_dev *dev, int cmd, void *data,
+		    int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	switch (cmd) {
+	case COMP_CMD_SET_VALUE:
+		return gapp_ctrl_set_val(dev, cdata);
+	case COMP_CMD_GET_VALUE:
+		return gapp_ctrl_get_val(dev, cdata, max_data_size);
+	case COMP_CMD_SET_DATA:
+		return gapp_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return gapp_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		comp_err(dev, "gapp_cmd(): unhandled command %d", cmd);
+		return -EINVAL;
+	}
+}
+
+static struct comp_dev *gapp_new(const struct comp_driver *drv,
+				 struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *gapp;
+	struct sof_ipc_comp_process *ipc_gapp =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_data *cd;
+	int ret = 0;
+
+	comp_cl_info(drv, "gapp_new()");
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		goto fail;
+
+	gapp = COMP_GET_IPC(dev, sof_ipc_comp_process);
+	ret = memcpy_s(gapp, sizeof(*gapp), ipc_gapp,
+		       sizeof(struct sof_ipc_comp_process));
+	if (ret)
+		goto fail;
+
+	if (ipc_gapp->size == sizeof(cd->config)) {
+		memcpy_s(cd->config, sizeof(cd->config), ipc_gapp->data,
+			 ipc_gapp->size);
+	}
+
+	cd->tuning_handler = comp_data_blob_handler_new(dev);
+	if (!cd->tuning_handler)
+		goto fail;
+
+	cd->state = GoogleAudioPostProcessingCreate();
+	if (!cd->state)
+		goto fail_internal;
+
+	comp_set_drvdata(dev, cd);
+
+	dev->state = COMP_STATE_READY;
+
+	comp_dbg(dev, "GAPP created c[0] %d c[1] %d", cd->config[0],
+		 cd->config[1]);
+
+	return dev;
+fail_internal:
+	comp_data_blob_handler_free(cd->tuning_handler);
+fail:
+	rfree(cd);
+	rfree(dev);
+	return NULL;
+}
+
+static void gapp_delete(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "gapp_delete()");
+
+	GoogleAudioPostProcessingDelete(cd->state);
+	comp_data_blob_handler_free(cd->tuning_handler);
+	rfree(cd);
+	rfree(dev);
+}
+
+static int gapp_params(struct comp_dev *dev,
+		       struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "gapp_params(): comp_verify_params failed.");
+		return ret;
+	}
+
+	return 0;
```

Comment:
might be simpler to just return comp_verify_params() directly. An error is printed in comp_verify_params() already.

Reply:
makes sense, updated.